Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs

```diff
@@ -0,0 +1,146 @@
+            // The component isn't doing anything anymore, so remove it
+            RemComp<NameModifierComponent>(entity);
+            return;
+        }
+        // We have at least one modifier, so we need to apply it to the entity.
+
+        // Get the final name with modifiers applied
+        var modifiedName = modifierEvent.GetModifiedName();
+
+        // Add the component if needed, and initialize it with the base name
+        if (!EnsureComp<NameModifierComponent>(entity, out var comp))
+            SetBaseName((entity, comp), meta.EntityName);
+
+        // Set the entity's name with modifiers applied
+        _metaData.SetEntityName(entity, modifiedName, meta, raiseEvents: false);
+    }
+}
+
+/// <summary>
+/// Raised on an entity when <see cref="NameModifierSystem.RefreshNameModifiers"/> is called.
+/// Subscribe to this event and use its methods to add modifiers to the entity's name.
+/// </summary>
+[ByRefEvent]
+public sealed class RefreshNameModifiersEvent : IInventoryRelayEvent
+{
+    /// <summary>
+    /// The entity's name without any modifiers applied.
+    /// If you want to base a modifier on the entity's name, use
+    /// this so you don't include other modifiers.
+    /// </summary>
+    public readonly string BaseName;
+
+    private readonly List<(LocId LocId, int Priority, List<(string, object)>? ExtraArgs)> _modifiers = [];
+
+    /// <inheritdoc/>
+    public SlotFlags TargetSlots => ~SlotFlags.POCKET;
+
+    /// <summary>
+    /// How many modifiers have been added to this event.
+    /// </summary>
+    public int ModifierCount => _modifiers.Count;
+
+    public RefreshNameModifiersEvent(string baseName)
+    {
+        BaseName = baseName;
+    }
+
+    /// <summary>
+    /// Adds a modifier to the entity's name.
+    /// The original name will be passed to Fluent as <c>$baseName</c> along with any <paramref name="extraArgs"/>.
+    /// Modifiers with a higher <paramref name="priority"/> will be applied later.
+    /// </summary>
+    public void AddModifier(LocId locId, int priority = 0, params (string, object)[]? extraArgs)
+    {
+        _modifiers.Add((locId, priority, extraArgs?.ToList()));
+    }
+
+    /// <summary>
+    /// Returns the final name with all modifiers applied.
+    /// </summary>
+    public string GetModifiedName()
+    {
+        // Start out with the entity's name name
+        var name = BaseName;
+
+        // Iterate through all the modifiers in priority order
+        foreach (var modifier in _modifiers.OrderBy(n => n.Priority))
+        {
+            // Grab any extra args needed by the Loc string
+            var args = modifier.ExtraArgs ?? [];
+            // Add the current version of the entity name as an arg
+            args.Add(("baseName", name));
+            // Resolve the Loc string and use the result as the base in the next iteration.
+            name = Loc.GetString(modifier.LocId, args.ToArray());
```

Comment:
I see what you meant with more allocations, but cant you make the _modifiers.ExtraArgs be an array and do something like 
```c#
        foreach (var (locId, _, extraArgs) in _modifiers.OrderBy(n => n.Priority))
        {
            // Grab any extra args needed by the Loc string
            var args = modifier.ExtraArgs ?? [];
            // Add the current version of the entity name as an arg
            Array.Resize(ref args, args.Length + 1);
            // Resolve the Loc string and use the result as the base in the next iteration.
            name = Loc.GetString(modifier.LocId, args);
```
you are still making a new array to copy but at least we get rid of 2 linq calls, just something i thought about, could just not work